tla-bmcmt/src/test/scala/at/forsyte/apalache/tla/bmcmt/CrossTestEncodings.scala

```diff
@@ -0,0 +1,245 @@
+ *   - There is no counterexample. There is a bug in one of the encodings.
+ */
+
+trait CrossTestEncodings extends AnyFunSuite {
+  // Override [[AnyFunSuite.withFixture]] to set the encodings.
+  protected var oracleEncoding: SMTEncoding = _
+  protected var verifierEncoding: SMTEncoding = _
+
+  /**
+   * Rewrite elem \in Set ~~> \E elem$selected \in Set: elem' := elem$selected.
+   */
+  private def rewriteElemInSet(elem: NameEx, set: TlaEx): TlaEx = {
+    val types = Map("e" -> elem.typeTag.asTlaType1(), "b" -> BoolT1())
+    val selected = tla.name(elem.name + "$selected").typed(types, "e")
+    tla
+      .apalacheSkolem(tla.exists(selected, set, tla.assign(tla.prime(elem) ? "e", selected) ? "b") ? "b")
+      .typed(types, "b")
+  }
+
+  /**
+   * Facade for running bounded model-checking through [[SeqModelChecker]].
+   *
+   * @param moduleName
+   *   name of the TLA+ module to check
+   * @param variableDecls
+   *   TLA+ variable declarations
+   * @param initTrans
+   *   init transition, see examples in [[getWitness]] and [[verify]]
+   * @param nextTrans
+   *   next transition, see examples in [[getWitness]] and [[verify]]
+   * @param invExpr
+   *   invariant to check
+   * @param smtEncoding
+   *   SMT encoding to use inside the model checker
+   * @return
+   *   a counterexample returned as VarNames -> Expressions binding in the initial state, i.e., after executing
+   *   [[initTrans]].
+   */
+  private def modelCheckAndGetCex(
+      moduleName: String,
+      variableDecls: Seq[TlaDecl],
+      initTrans: List[TlaEx],
+      nextTrans: List[TlaEx],
+      invExpr: TlaEx,
+      smtEncoding: SMTEncoding): Map[String, TlaEx] = {
+    // prepare the invariant
+    val invDecl = tla.declOp("Inv", invExpr).as(OperT1(Nil, BoolT1()))
+    val decls = variableDecls :+ invDecl
+    val invariantsAndNegations = List((invExpr, tla.not(invExpr).as(BoolT1())))
+    val verificationConditions = CheckerInputVC(invariantsAndNegations)
+
+    val module = TlaModule(moduleName, decls)
+
+    val checkerInput = new CheckerInput(module, initTrans, nextTrans, None, verificationConditions)
+    val checkerParams = new ModelCheckerParams(checkerInput, 0)
+
+    val solverContext = RecordingSolverContext.createZ3(None, SolverConfig.default)
+    val rewriter: SymbStateRewriterImpl = smtEncoding match {
+      case `oopsla19Encoding` => new SymbStateRewriterImpl(solverContext)
+      case `arraysEncoding`   => new SymbStateRewriterImplWithArrays(solverContext)
+    }
+
+    val ctx = new IncrementalExecutionContext(rewriter)
+    val trex = new TransitionExecutorImpl(checkerParams.consts, checkerParams.vars, ctx)
+
+    // run the model checker with listener
+    val listener = new CollectCounterexamplesModelCheckerListener()
+    val checker = new SeqModelChecker(checkerParams, checkerInput, trex, Seq(listener))
+
+    // check the outcome
+    val outcome = checker.run()
+    assert(Error(1) == outcome)
```

Comment:
Is it always the case that we should expect a counterexample? This is probably not true for the verifier.

Reply:
In principle, there should always be a counterexample, since `Init == ... /\ found \in BOOLEAN` and `NotFound == ~found`. Unless we put unsatisfiable constraints in `Init` and the model checker reports `Deadlock` because it cannot take the init transition. But this we should anyhow avoid.